File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/x/XDetector.java
package hu.kdea.szavazas.ballotprocessor.x;

import boofcv.alg.filter.binary.BinaryImageOps;
import boofcv.struct.image.GrayU8;
import hu.kdea.szavazas.ballotprocessor.GridConstants;
import hu.kdea.szavazas.ballotprocessor.common.PointData;
import hu.kdea.szavazas.ballotprocessor.common.RectangleData;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;

public class XDetector {
    @Inject
    public XDetector() {
    }

    public boolean detect(GrayU8 binary, RectangleData outerRect) {
        return detectWithDebug(binary, outerRect).detected();
    }

    public XDetectionResultData detectWithDebug(GrayU8 binary, RectangleData outerRect) {
        RectangleData innerRect = computeInnerRect(outerRect);
        if (innerRect.width() <= 0 || innerRect.height() <= 0) {
            return new XDetectionResultData(false, null);
        }
        GrayU8 originalCell = cropCell(binary, innerRect);
        ErosionResultData erosionResult = maybeErode(originalCell);
        GrayU8 workMat = erosionResult.processed();
        GrayU8 erodedCell = erosionResult.eroded();
        GrayU8 skeleton = BinaryImageOps.thin(workMat, -1, null);
        List<PointData> branchPoints = findBranchPoints(skeleton);
        boolean hasX = branchPoints.size() >= GridConstants.MIN_BRANCHES;
        CellDebugData debug = new CellDebugData(outerRect, innerRect, originalCell, erodedCell, skeleton, branchPoints, hasX);
        return new XDetectionResultData(hasX, debug);
    }

    private RectangleData computeInnerRect(RectangleData outer) {
        return new RectangleData(
            outer.x() + GridConstants.X_MARGIN,
            outer.y() + GridConstants.X_MARGIN,
            outer.width() - 2 * GridConstants.X_MARGIN,
            outer.height() - 2 * GridConstants.X_MARGIN
        );
    }

    private GrayU8 cropCell(GrayU8 source, RectangleData rect) {
        GrayU8 out = new GrayU8(rect.width(), rect.height());
        for (int y = 0; y < rect.height(); y++) {
            for (int x = 0; x < rect.width(); x++) {
                out.set(x, y, source.get(rect.x() + x, rect.y() + y));
            }
        }
        return out;
    }

    private ErosionResultData maybeErode(GrayU8 cell) {
        if (GridConstants.ERODE_KERNEL_SIZE <= 0 || GridConstants.ERODE_ITERATIONS <= 0) {
            return new ErosionResultData(cell, null);
        }
        GrayU8 eroded = BinaryImageOps.erode8(cell, GridConstants.ERODE_ITERATIONS, null);
        return new ErosionResultData(eroded, eroded);
    }

    private List<PointData> findBranchPoints(GrayU8 skeleton) {
        List<PointData> points = new ArrayList<>();
        for (int y = 1; y < skeleton.height - 1; y++) {
            for (int x = 1; x < skeleton.width - 1; x++) {
                if (skeleton.get(x, y) != 0 && neighbourCount(skeleton, x, y) >= 3) {
                    points.add(new PointData(x, y));
                }
            }
        }
        return points;
    }

    private int neighbourCount(GrayU8 skeleton, int x, int y) {
        int count = 0;
        for (int dy = -1; dy <= 1; dy++) {
            for (int dx = -1; dx <= 1; dx++) {
                if (dx == 0 && dy == 0) {
                    continue;
                }
                if (skeleton.get(x + dx, y + dy) != 0) {
                    count++;
                }
            }
        }
        return count;
    }
}

File: test-utils/src/main/java/hu/kdea/szavazas/ballotprocessor/x/XDetector.kt
package hu.kdea.szavazas.ballotprocessor.x

import boofcv.alg.filter.binary.BinaryImageOps
import boofcv.struct.image.GrayU8
import hu.kdea.szavazas.ballotprocessor.GridConstants
import hu.kdea.szavazas.ballotprocessor.Logger
import hu.kdea.szavazas.ballotprocessor.common.Point
import hu.kdea.szavazas.ballotprocessor.common.Rect

class XDetector {

    fun detect(binary: GrayU8, outerRect: Rect): Boolean =
        detectWithDebug(binary, outerRect).first

    fun detectWithDebug(binary: GrayU8, outerRect: Rect): Pair<Boolean, CellDebugData?> {
        val innerRect = computeInnerRect(outerRect)
        if (innerRect.width <= 0 || innerRect.height <= 0) return false to null

        val originalCell = cropCell(binary, innerRect)
        val (workMat, erodedCell) = maybeErode(originalCell)
        val skeleton = BinaryImageOps.thin(workMat, -1, null)
        val branchPoints = findBranchPoints(skeleton)
        val hasX = branchPoints.size >= GridConstants.MIN_BRANCHES

        Logger.d(
            "XDetector",
            "Cell $outerRect: ${branchPoints.size} branch point(s) -> ${if (hasX) "X" else "no X"}"
        )

        val debug = CellDebugData(outerRect, innerRect, originalCell, erodedCell, skeleton, branchPoints)
        return hasX to debug
    }

    private fun computeInnerRect(outer: Rect): Rect = Rect(
        outer.x + GridConstants.X_MARGIN,
        outer.y + GridConstants.X_MARGIN,
        outer.width - 2 * GridConstants.X_MARGIN,
        outer.height - 2 * GridConstants.X_MARGIN
    )

    private fun cropCell(source: GrayU8, rect: Rect): GrayU8 {
        val out = GrayU8(rect.width, rect.height)
        for (y in 0 until rect.height) {
            for (x in 0 until rect.width) {
                out.set(x, y, source.get(rect.x + x, rect.y + y))
            }
        }
        return out
    }

    private fun maybeErode(cell: GrayU8): Pair<GrayU8, GrayU8?> {
        if (GridConstants.ERODE_KERNEL_SIZE <= 0 || GridConstants.ERODE_ITERATIONS <= 0)
            return cell to null
        val eroded = BinaryImageOps.erode8(cell, GridConstants.ERODE_ITERATIONS, null)
        return eroded to eroded
    }

    private fun findBranchPoints(skel: GrayU8): List<Point> {
        val points = mutableListOf<Point>()
        for (y in 1 until skel.height - 1) {
            for (x in 1 until skel.width - 1) {
                if (skel.get(x, y) != 0 && neighbourCount(skel, x, y) >= 3) {
                    points.add(Point(x, y))
                }
            }
        }
        return points
    }

    private fun neighbourCount(img: GrayU8, x: Int, y: Int): Int {
        var cnt = 0
        for (dy in -1..1) for (dx in -1..1) {
            if (dx == 0 && dy == 0) continue
            if (img.get(x + dx, y + dy) != 0) cnt++
        }
        return cnt
    }
}